### PR TITLE
Add PyAlbany to the Albany recipe

### PR DESCRIPTION
**Pull request by kliegeois**
@xylar @ikalash 

This PR adds a new option so to false by default that allows the user to build PyAlbany using the spack recipe.

### CODE REVIEW

**Comment by kliegeois** [comment]
With the second commit, it is now possible to correctly run PyAlbany examples.
The unit tests are not yet working but the modification will be on the Albany side now.
I will remove the `WIP`.

**Review by xylar — APPROVED** [review]
To me, this look great!  @ikalash, feel free to merge this once your happy.  Maybe do some testing since you're presumably familiar with how to make use of the python interface, whereas I'm not.

**Comment by kliegeois** [comment]
Thanks for the review @xylar !

**Comment by kliegeois** [comment]
Actually, I will need to add a scipy dependency too (I forgot that some of the ctest tests use it). I will update the PR asap.

**Comment by ikalash** [comment]
@kliegeois : let me know when this is ready to test.  A follow-up question I had was whether we want to modify the nightly spack build I have to enable pyalbany, once your changes are merged.  Probably the answer is yes?

File: var/spack/repos/builtin/packages/albany/package.py
Comment (xylar) [inline var/spack/repos/builtin/packages/albany/package.py]: After I rebase onto v0.19.0, we're going to have to rebase this and change all `'` to `"` to be consistent with v0.19.0's formatting convention.

**Comment by kliegeois** [comment]
I have resolved the merge conflict locally, I am giving a try to the build. After that, I will update the PR and let you know when you can try it. Thanks!

**Comment by kliegeois** [comment]
This branch is now conflict-free @xylar.
Along with https://github.com/sandialabs/Albany/commit/f38b8f00eb0209c86a7c813810abe281dd0454e6 you can build PyAlbany and run the ctest tests.

Once Albany is compiled and install with spack, do the following commands to test PyAlbany tests:

```
export PYTHONPATH=""
spack load albany
spack cd albany
cd ../spack-build-... #this needs to be updated based on the name of the build directory
ctest -R PyAlbany
```

@ikalash if you have an easy way to `cd ../spack-build-...` we can use this as a way to test PyAlbany in the nightly tests. That would be great! Thanks.

**Comment by ikalash** [comment]
Thanks @kliegeois .  Does something special need to be done to turn on PyAlbany?  E.g., build albany+pyalbany similar to how we turn on MPAS in Albany (albany+mpas)?   
 I would maybe vote to add an option like this so that the PyAlbany tests run with just ctest, similar to how it works in Albany built using cmake.  Thoughts?  I already cd to the build directory where Albany gets built in my nightlies to run ctest, so that part is not a problem.

**Comment by kliegeois** [comment]
The option is "+py" and is off by default.

**Comment by xylar** [comment]
@kliegeois, I'm still seeing a bunch of commits we don't want to see.  Can you try the rebase again, removing any commits you didn't make?  (I tried changing the base and changing it back to see if that helped but it didn't seem to.)

**Comment by xylar** [comment]
@kliegeois, it looks like you did a merge to handle the conflicts but that isn't going to work in this situation.  We need a rebase because we don't want commits from the original `E3SM-Project/spack/develop` before I rebased that onto `v0.19.0`.  I can do the rebase for you if you need.

**Comment by xylar** [comment]
Even easier, just hard-reset to `E3SM-Project/spack/develop` and then cherry-pick 152e7700b0af52769104405a1699e0703688b19f, since that's the only new commit.

**Comment by kliegeois** [comment]
Done @xylar :)

**Review by xylar — APPROVED** [review]
Great, thanks!

**Comment by ikalash** [comment]
> ctest -R PyAlbany

Why does one need to do ctest -R PyAlbany if PyAlbany is turned on with the +py option?  Wouldn't the tests run with just normal ctest if the code was built with +py?

FYI, I am traveling this week for a project review, and will not get a chance to test this until at least Thursday, possibly not until the weekend.  My apologies for the delay.

**Comment by kliegeois** [comment]
> > ctest -R PyAlbany
> 
> Why does one need to do ctest -R PyAlbany if PyAlbany is turned on with the +py option? Wouldn't the tests run with just normal ctest if the code was built with +py?
> 
> FYI, I am traveling this week for a project review, and will not get a chance to test this until at least Thursday, possibly not until the weekend. My apologies for the delay.

@ikalash just not to run all the other tests. However, for the nightly build, we want to run them all (_and in the darkness bind them_   :mage_man:).

**Comment by xylar** [comment]
@ikalash, merge this whenever you're happy.

**Comment by ikalash** [comment]
Thanks for the reply @kliegeois .  @xylar : will do.

**Comment by ikalash** [comment]
I have tested this - sorry again about the delay.  It looks good except the following test is failing: 

```
The following tests FAILED:
	170 - BlockDiscretization_Jacobian_Unit_Test (Failed)
Errors while running CTest
Output from these tests are in: /tmp/ikalash/spack-stage/spack-stage-albany-develop-6bkhva4eddgtbxn2n2m4wew444slo6fa/spack-build-6bkhva4/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.

```
Do you know what it's about?  It seems to not have anything to do with PyAlbany.  Tagging @bartgol and @mperego, in case they have an idea.

**Comment by kliegeois** [comment]
@ikalash this test is tested because it requires python with scipy and numpy to run. It tests 2 discretization strategies of Albany and then compare the matrices in python (some reordering of the indices is done in a python script for the unit test if I recall). The test was not run before with spack as there was no python available. But that does not explain why the test fails though.

**Comment by ikalash** [comment]
I looked at the verbose output and the issue is numpy is not found: 

```
test 1
    Start 1: BlockDiscretization_Jacobian_Unit_Test

1: Test command: /home/ikalash/spack/opt/spack/linux-rhel7-westmere/gcc-10.1.0/python-3.10.8-bye6aeb47a5bxszi44dvys557ovaxjpq/bin/python3 "/tmp/ikalash/spack-stage/spack-stage-albany-develop-6bkhva4eddgtbxn2n2m4wew444slo6fa/spack-build-6bkhva4/tests/unit/disc/matrices_comparison.py"
1: Working Directory: /tmp/ikalash/spack-stage/spack-stage-albany-develop-6bkhva4eddgtbxn2n2m4wew444slo6fa/spack-build-6bkhva4/tests/unit/disc
1: Test timeout computed to be: 10000000
1: Traceback (most recent call last):
1:   File "/tmp/ikalash/spack-stage/spack-stage-albany-develop-6bkhva4eddgtbxn2n2m4wew444slo6fa/spack-build-6bkhva4/tests/unit/disc/matrices_comparison.py", line 1, in <module>
1:     import numpy as np
1: ModuleNotFoundError: No module named 'numpy'
1/3 Test #1: BlockDiscretization_Jacobian_Unit_Test ...***Failed    0.12 sec
```
The python that is being used in the test is the one that is built by spack.  I don't know how to tell spack python to install package like numpy - @xylar do you know?  In any case, I would think the correct behavior of this test would be to only run if python has numpy.  @kliegeois would it be possible to make this change?  I will go ahead and approve/merge this PR, b/c the changes we are discussing would likely go in Albany not spack.

**Comment by kliegeois** [comment]
@ikalash the python that spack install has numpy but I am almost sure the issue is in the PYTHONPATH of the test.

**Comment by ikalash** [comment]
@kliegeois .  Ok.  how do you propose to troubleshoot / fix this?

**Review by ikalash — APPROVED** [review]

**Comment by kliegeois** [comment]
@ikalash if the issue is what I think it is, it will be an easy cmake fix on the Albany side. However, I can't do it now. I might be able to do it this weekend. If not, I will do it early on Monday. Does that work?

**Comment by ikalash** [comment]
Ah ok, thanks for clarifying.  Oh yes, Monday is just fine - no need to work on this over the weekend!  Thanks in advance!

**Comment by kliegeois** [comment]
@ikalash I can reproduce the issue. I have been able to solve it but there is a segfault when the test is actually run (after finding numpy). I am trying to see if this can be reproduce with our standard build and I may have to open an issue on the Albany repo.

**Comment by kliegeois** [comment]
@ikalash it is resolved and the segfault was a linking issue on my spack build that other unit tests have too (those tests were passing on your build so it is a local issue on my side).

Let me know if the test is still failing.

**Comment by ikalash** [comment]
Thanks @kliegeois .  Am I correct to understand that you made the change to Albany master, and there is no need to review / approve / merge in any PR?  If so, I can go ahead and test the spack build.

**Comment by kliegeois** [comment]
Yes, that's correct @ikalash (the fix was one line https://github.com/sandialabs/Albany/commit/dab3dc171a1b0479a14d6f2eb675c45301755d47 )

**Comment by ikalash** [comment]
I tested things and everything looks good.  Thanks for fixing it so fast @kliegeois !

**Pull request merged**
